### PR TITLE
fix(gateguard): look past sudo/doas/env when classifying destructive commands

### PR DESCRIPTION
**Pull request by ntdat812**
## What Changed

Every token-based detector in `gateguard-fact-force.js` keys on `tokens[0]`, so a privilege or environment wrapper hides the command being run. `stripCommandWrappers()` now removes leading wrapper commands, their flags, and `VAR=value` assignment prefixes before the detectors see the tokens — applied in the per-segment loop (`isDestructiveRm`, `isDestructiveGit`) and in `isDestructiveFindExec`.

I found this while fixing #2642; it is a separate defect from the `dd` one, so it is a separate PR.

## Why This Change

Measured against `main` with a **primed session** (the first Bash command of a session trips a separate routine gate that otherwise masks this), driven through the published hook runner:

| command | main | this PR |
| --- | --- | --- |
| `rm -rf /important/data` | denied | denied |
| `sudo rm -rf /important/data` | **allowed** | denied |
| `sudo -u root rm -rf /data` | **allowed** | denied |
| `sudo --user=root rm -rf /data` | **allowed** | denied |
| `doas rm -rf /data` | **allowed** | denied |
| `env rm -rf /data` | **allowed** | denied |
| `FOO=bar rm -rf /data` | **allowed** | denied |
| `command rm -rf /data` | **allowed** | denied |
| `nohup rm -rf /data` | **allowed** | denied |
| `sudo git reset --hard` | **allowed** | denied |
| `sudo git clean -fd` | **allowed** | denied |
| `sudo git push --force` | **allowed** | denied |
| `sudo find . -exec rm {} ;` | **allowed** | denied |
| `echo a && sudo rm -rf /data` | **allowed** | denied |

The gate missed exactly the spellings it matters most for: the privileged ones. Adding four characters to a command was enough to turn a hard deny into a pass.

`sudo truncate table t` was already denied — the SQL arm is a text match and never looked at a command word at all, which is why the gap only shows up in the token-based detectors.

## How it is bounded

Widening a *deny* gate risks blocking legitimate work, so two limits are deliberate:

- **Flags are skipped only after a wrapper has been seen.** Without that, the scan could walk into an unwrapped command's own arguments.
- **Wrapper flags that consume a following token are listed** (`-u`, `-g`, `-p`, `-C`, `-h`, `-r`, `-t`, `-U`, and long forms), so `sudo -u root rm -rf /` resolves to `rm` and not to `root`. `--flag=value` needs no entry — it is one token.

Verified still **allowed**: `sudo apt-get update`, `sudo systemctl restart nginx`, `sudo ls -la /etc`, `command ls`, `env NODE_ENV=production npm run build`.

## Testing Done

- [x] Manual testing completed
- [ ] Automated tests pass locally (`node tests/run-all.js`) — see the note below
- [x] Edge cases considered and tested

13 new deny cases and 5 new allow cases in `tests/hooks/gateguard-fact-force.test.js`, driven end-to-end through `run-with-flags.js`.

Reverting only `scripts/hooks/gateguard-fact-force.js`:

```
✗ denies a wrapper-prefixed destructive command: sudo rm -rf /important/data
✗ … sudo -u root rm -rf /important/data
✗ … sudo --user=root rm -rf /important/data
✗ … doas rm -rf /important/data
✗ … env rm -rf /important/data
✗ … FOO=bar rm -rf /important/data
✗ … command rm -rf /important/data
✗ … nohup rm -rf /important/data
✗ … sudo git reset --hard
✗ … sudo git clean -fd
✗ … sudo git push --force
✗ … sudo find . -exec rm {} ;
✗ … echo hello && sudo rm -rf /important/data
149 passed, 13 failed
```

With the change: **162 passed, 0 failed**. The 5 must-allow cases pass either way by design — they are the over-blocking guard, not evidence of the bug.

**Why `run-all.js` is unchecked:** it stops in my environment at `tests/lib/claude-plugin-setup.test.js`, which performs real marketplace `git clone`s (`fatal: fetch-pack: invalid index-pack output`). That is environmental and unrelated. `tests/hooks/gateguard-fact-force.test.js` is the only test file that references gateguard (`grep -rl gateguard tests/`), so the suite I did run covers this change's blast radius.

## Type of Change

- [x] `fix:` Bug fix

## Security & Quality Checklist

- [x] No secrets or API keys committed
- [x] JSON files validate cleanly (none touched)
- [x] Shell scripts pass shellcheck (none touched)
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

## Relationship to #2829

Independent — this branches from `main` and touches different functions, so the two can merge in either order.

There is one small overlap to reconcile afterwards: #2829 adds `isDestructiveDd`, which carries its own inline wrapper-skipping loop. Once both are in, that loop should call `stripCommandWrappers()` instead. I will send that one-line follow-up, or fold it in here if you prefer to take this one first — your call. As a side effect it would also fix `sudo -u root dd if=…`, which #2829's inline loop does not handle (it stops at `root`).

### CODE REVIEW

**Comment by ntdat812** [comment]
All three findings are correct. I reproduced each one against the production hook before changing anything, and pushed a fix in `1e2aa07`.

## What was still open

Driven through `run-with-flags.js` with a primed session, on the previous head of this branch:

| command | before | now |
| --- | --- | --- |
| `find . -exec sudo rm {} ;` | **allowed** | denied |
| `find . -exec doas rm {} ;` | **allowed** | denied |
| `find . -exec sudo git reset --hard {} ;` | **allowed** | denied |
| `command -p rm -rf /data` | **allowed** | denied |
| `nice -n 10 rm -rf /data` | **allowed** | denied |
| `stdbuf -o L rm -rf /data` | **allowed** | denied |
| `stdbuf -oL rm -rf /data` | **allowed** | denied |
| `ionice -c 3 rm -rf /data` | **allowed** | denied |
| `sudo "rm" -rf /data` | **allowed** | denied |
| `sudo 'rm' -rf /data` | **allowed** | denied |
| `env "rm" -rf /data` | **allowed** | denied |
| `doas "rm" -rf /data` | **allowed** | denied |
| `sudo bash -c "rm -rf /data"` | **allowed** | denied |

For the record: all of these are **pre-existing on `main`**, not regressions introduced by the first pass — I confirmed by reverting only the hook source. That does not make them out of scope; they are the same defect class this PR claims to close, so they belong here.

## The fixes

**Option arity is per wrapper.** `COMMAND_WRAPPERS` is now a `Map` from wrapper name to that wrapper's *own* value-taking options. The shared set was wrong in both directions: it made `command -p` swallow `rm` (`-p` is boolean for `command`, a prompt string for `sudo`), and it left `nice -n 10` / `stdbuf -o L` / `ionice -c 3` stopping on the value token. Attached short forms (`stdbuf -oL`) and `--flag=value` need no entries — they are a single token that gets skipped either way.

**`find -exec`.** The executed command now goes through `stripCommandWrappers()` before `baseCmd` and `findGitSubcommand()` are derived, so `-exec sudo rm` is classified the same as `-exec rm`.

**Quote-aware pass.** Each `quoteAwareSegments()` segment is normalized before the `rm`/git/`find -exec` detectors run. As a side effect this also closes `sudo bash -c "rm -rf /"` — the `sh -c` recursion keys on `tokens[0]`, which the wrapper was hiding.

## Verification

`node tests/hooks/gateguard-fact-force.test.js`: **182 passed, 0 failed** (was 162 — 13 new deny cases, 7 new allow cases).

Reverting only `scripts/hooks/gateguard-fact-force.js` to `d8409a4`: **156 passed, 26 failed**, all 26 being the wrapper deny cases.

Widening a deny gate risks blocking real work, so the allow side was measured too. Still allowed: `sudo apt-get update`, `sudo systemctl restart nginx`, `sudo ls -la /etc`, `sudo -u root ls -la /etc`, `sudo -u postgres psql -c "select 1"`, `env NODE_ENV=production npm run build`, `command ls`, `command -p ls`, `command -v rm`, `nice -n 10 npm run build`, `ionice -c 3 npm test`, `stdbuf -o L npm test`, `find . -exec sudo chmod 644 {} ;`, `find . -name "*.log" -exec ls {} ;`.

## Not addressed

`stripCommandWrappers()` operates on tokens, so a wrapper reached through a variable (`W=sudo; $W rm -rf /`) or a shell function alias is still out of reach. That is a different layer than this PR touches and I would rather not widen a deny gate speculatively — happy to open it separately if you want it covered.

File: scripts/hooks/gateguard-fact-force.js [hunk 1]
```diff
@@ -375,6 +379,86 @@ function commandBasename(token) {
...
+    'sudo',
+    new Set([
+      '-a', '-C', '-c', '-D', '-g', '-h', '-p', '-R', '-r', '-T', '-t', '-U', '-u',
+      '--auth-type', '--close-from', '--login-class', '--chdir', '--group',
+      '--host', '--prompt', '--chroot', '--role', '--command-timeout', '--type',
+      '--other-user', '--user',
+    ]),
+  ],
+  ['doas', new Set(['-a', '-C', '-u'])],
+  ['env', new Set(['-u', '-C', '-S', '--unset', '--chdir', '--split-string'])],
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply scripts/hooks/gateguard-fact-force.js:471]: Confirmed and fixed in 89c1bf4f.

Reproduced against the built hook first — all three spellings were allowed while the bare command was denied:

```
deny    rm -rf /important/data
(none)  env -S "rm -rf /important/data"
(none)  env --split-string="rm -rf /important/data"
(none)  env -S"rm -rf /important/data"
```

The cause was in this table, not the detectors: `env` listed `-S` / `--split-string` as value-consuming, so `stripCommandWrappers` skipped the option *and* the string — and the string is the command.

It is now expanded into tokens and the walk continues, so a wrapper nested inside it is stripped too (`env -S "sudo rm -rf /x"` denies). Bounded by `MAX_WRAPPER_EXPANSIONS` so a self-referential string cannot spin. `env -S "npm run build"` stays allowed.

192 passed, 0 failed. Removing the expansion fails 5 of them.

File: scripts/hooks/gateguard-fact-force.js [hunk 2]
```diff
@@ -375,6 +379,86 @@ function commandBasename(token) {
...
+  ['command', new Set()],
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply scripts/hooks/gateguard-fact-force.js:472]: Agreed, and fixed in 89c1bf4f — thanks, this one was a false positive in the opposite direction.

`command -v git reset --hard` was denied before; it now returns no executable tokens, along with `-V` and combined forms like `-pv`. `-p` alone still executes its operand, so `command -p rm -rf /important/data` continues to deny.

Note `command -v rm` passed before only by accident — `rm` on its own is not destructive without `-rf`, so the lookup was never actually exercised. Both spellings are now covered by explicit tests.

Removing the lookup guard fails 3 tests.

File: scripts/hooks/gateguard-fact-force.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ntdat812) [reply scripts/hooks/gateguard-fact-force.js:471]: Same finding as the CodeRabbit thread above, and it reproduces exactly as you describe — fixed in 89c1bf4f. Both `-S "..."`, `--split-string=...`, and the attached `-S"..."` form were allowed; all three now deny, including when the split string itself starts with another wrapper.

File: scripts/hooks/gateguard-fact-force.js [hunk 3]
```diff
@@ -375,6 +379,147 @@ function commandBasename(token) {
...
+function envSplitString(token, next) {
+  if (token === '-S' || token === '--split-string') {
+    return typeof next === 'string' ? { value: next, consumed: 2 } : null;
+  }
+  if (token.startsWith('--split-string=')) {
+    return { value: token.slice('--split-string='.length), consumed: 1 };
+  }
+  if (token.startsWith('-S') && token.length > 2) {
+    return { value: token.slice(2), consumed: 1 };
+  }
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply scripts/hooks/gateguard-fact-force.js:516]: Confirmed and fixed in f48abf7f. `envSplitString` anchored `-S` at position 1; getopt gives the first argument-taking option in a cluster the rest of the token, so `-S` is reachable through `-v`/`-i`, attached or separate. It now walks the cluster.

The walk has to stop at `-u`/`-C`, which claim the cluster themselves — and that turned out to matter more than the reported case. `env -uSHELL rm -rf /important/data` is `-u` unsetting SHELL followed by a real deletion; without the stop, `HELL` is read as the split string, becomes the command word, and the deletion is allowed:

```
with the stop:     deny    env -uSHELL rm -rf /important/data
without the stop:  (none)  env -uSHELL rm -rf /important/data
```

I only found that because removing the stop passed all 200 tests, so it is now pinned by its own case.

```
deny    env -vS"rm -rf /important/data"
deny    env -vS "rm -rf /important/data"
deny    env -iS "rm -rf /important/data"
(none)  env -uS FOO
(none)  env -vS "npm test"
```

202 passed, 0 failed; anchoring `-S` at position 1 again fails 3 of them.

File: scripts/hooks/gateguard-fact-force.js [hunk 4]
```diff
@@ -375,6 +379,147 @@ function commandBasename(token) {
...
+      continue;
+    }
+    if (valueFlags && token.startsWith('-')) {
+      // `env -S "rm -rf /"` RUNS that string. Consuming it as an opaque option
+      // value hid the whole command behind the option, so expand it and keep
+      // walking. Bounded so a self-referential string cannot spin.
+      if (wrapper === 'env' && expansions < MAX_WRAPPER_EXPANSIONS) {
+        const split = envSplitString(token, rest[index + 1]);
+        if (split) {
+          rest = tokenize(split.value).concat(rest.slice(index + split.consumed));
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply scripts/hooks/gateguard-fact-force.js]: Confirmed and fixed in f48abf7f — you are right that the whitespace-only `tokenize()` was the wrong tool for a value that carries its own quoting. It now expands with the file's quote-aware `tokenizeAllowlistedShellWords()`, falling back to the plain split if that cannot parse.

Worth noting the single-level form had the same hole, not just the nested one:

```
deny    env -S '"rm" -rf /important/data'
deny    env -S 'env -S "rm" -rf /important/data'
```

Both are covered by regression tests now. Dropping the quote awareness fails 2 of them.

File: scripts/hooks/gateguard-fact-force.js [hunk 5]
```diff
@@ -643,7 +807,14 @@ function isDestructiveFindExec(command) {
...
+  // executed command needs the same wrapper normalization as a top-level one.
+  const execCommand = stripCommandWrappers(execTokens);
+  if (execCommand.length === 0) {
+    return false;
+  }
+
+  const baseCmd = commandBasename(execCommand[0]);
 
   // Directly destructive commands inside -exec
   if (baseCmd === 'rmdir' || baseCmd === 'unlink') {
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply scripts/hooks/gateguard-fact-force.js:915]: Correct, and reproduced before changing anything. Against the built hook, `find . -exec sh -c "rm -rf {}" \;` was ALLOW while `find . -exec rm {} \;` and a bare `sh -c "rm -rf ..."` were both DENY — the wrapper ended the inspection exactly where the destructive command began.

Fixing it turned up a second gap on the way. `isDestructiveQuoteAware` already recurses into `sh -c`, so it should have caught this via its own `isDestructiveFindExec` call, but it handed that check a re-joined string — which flattened the quoted payload back into separate words. It now passes the tokens it already holds, through a token-taking half of the same function.

The other half was `-c` being matched only as a bare token. Short options cluster, and I checked the spellings against the shell rather than assuming:

- `sh -ec 'echo RAN'`, `sh -xc`, `sh -ce` → all run the payload, so `sh -ec 'rm -rf /important/data'` was allowed **without `find` involved at all**.
- `sh -c'cmd'` → `invalid option`. The payload always comes from the next argument, never attached.
- `sh -oc 'cmd'` → `invalid option name`, and `sh -- -c 'cmd'` → `No such file or directory`. `-o`/`-O` take a value and `--` ends the options, so the scan stops at both; the shell does not run those strings, so stopping loses nothing.

Pushed in `d29d8a3`. Suite is 202 → 221, and mutation testing pinned each piece: re-joining the tokens loses all eight `-exec sh -c` spellings, and matching `-c` as a whole option body loses every clustered one. It also showed my first two allow-list cases were worthless — with a benign payload they passed with the arity and `--` stops removed. They now carry a destructive payload, which is the only thing that discriminates.

File: scripts/hooks/gateguard-fact-force.js [hunk 6]
```diff
@@ -333,6 +333,48 @@ function quoteAwareSegments(input) {
...
+    // either one there is no `-c` left to find.
+    if (token === '--' || token === '-' || !token.startsWith('-')) return null;
+    if (token.startsWith('--')) continue;
+
+    for (const flag of token.slice(1)) {
+      if (flag === 'c') {
+        const payload = tokens[i + 1];
+        return typeof payload === 'string' ? payload : null;
+      }
+      if (SHELL_SHORT_OPTIONS_WITH_VALUE.has(flag)) return null;
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply scripts/hooks/gateguard-fact-force.js]: Correct, and this one is a regression I introduced on this branch — worth stating plainly. Both spellings were denied at `f48abf7f` and allowed at `d29d8a3`:

```
                                             parent   after my change
sh -o errexit -c "rm -rf /important/data"     DENY  ->  ALLOW
bash --init-file /dev/null -c "rm -rf ..."    DENY  ->  ALLOW
```

The old `indexOf('-c')` found the flag wherever it sat. My cluster reader replaced that with a left-to-right scan, and I wrote "this option takes a value" as "stop looking" — but the option consumes exactly one value and the scan has to resume after it. Both of your probe commands do run their payload; I re-ran them here, including `bash --rcfile /dev/null -ic`, which prints its tty warning and then runs the string anyway.

Fixed in `788bb6a`. Only a `c` sharing a token with the value is excluded now — `sh -oc 'cmd'`, which the shell rejects with "invalid option name" rather than running it — and `--rcfile`/`--init-file` are skipped along with the token that follows them.

```
sh -o errexit -c "rm -rf /important/data"        DENY
bash --init-file /dev/null -c "rm -rf ..."       DENY
bash --rcfile /dev/null -ic "rm -rf ..."         DENY
find . -exec sh -o errexit -c "rm -rf {}" ;      DENY
sh -oc "rm -rf /important/data"                  ALLOW
sh -o errexit -c "npm test"                      ALLOW
```

All four denials are pinned by tests, and the two allowances carry a destructive payload so they actually discriminate.

File: scripts/hooks/gateguard-fact-force.js [hunk 7]
```diff
@@ -375,6 +416,166 @@ function commandBasename(token) {
...
+          // word never matches. Falls back to a plain split if parsing fails.
+          const words = tokenizeAllowlistedShellWords(split.value) || tokenize(split.value);
+          rest = words.concat(rest.slice(index + split.consumed));
+          index = 0;
+          valueFlags = null;
+          wrapper = null;
+          expansions += 1;
+          continue;
+        }
+      }
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply scripts/hooks/gateguard-fact-force.js]: Confirmed, and reachable with a command that really runs. `-Senv` is env's attached split-string form, so the layers nest for real — `env -Senv -S "echo RAN"` prints `RAN`, as does the eight-deep version.

Sweeping the depth against the built hook shows the edge exactly where you said it was:

```
 7 nested  DENY   env -Senv ... -S "rm -rf /important/data"
 8 nested  ALLOW
 9 nested  ALLOW
```

At the cap the walk fell through to the branch that treats `-S` as an opaque option value — which puts the command back behind the option, the precise bypass the expansion exists to close. So the cap did not merely stop early, it restored the original bug, and one extra wrapper bought a pass.

Raising the number only moves the edge, so `788bb6a` removes it as a semantic limit. Each expansion replaces the option with the words of a string that is a strict substring of the token it came from, so the remaining character count strictly decreases and the walk terminates on its own; the loop is now bounded by that character count, which keeps the termination guarantee while making the bound unreachable.

```
   0 nested  DENY   2ms
   8 nested  DENY   0ms
  50 nested  DENY   1ms
1000 nested  DENY  11ms
benign 200-deep -> ALLOW
```

Regression tests cover 8, 9, and 40 layers on the destructive side. A 40-layer benign one is in the allow matrix too — depth is not what makes a command destructive, and a bound that gave up by denying would have gated it. Suite is 221 → 230.

File: tests/hooks/gateguard-fact-force.test.js [hunk 8]
```diff
@@ -253,6 +253,174 @@ function runTests() {
...
+    'find . -exec sh -c "rm -rf {}" +',
+    'find . -exec sh -c "git reset --hard" ;',
+    'find . -exec sh -c "sudo rm -rf /important/data" ;',
+    // Short options cluster, so `-c` is reached through `-e`/`-x` as well —
+    // attached or not, and whether or not `c` comes first.
+    'sh -ec "rm -rf /important/data"',
+    'sh -xc "rm -rf /important/data"',
+    'sh -ce "rm -rf /important/data"',
+    'bash -ec "rm -rf /important/data"',
+    'find . -exec sh -ec "rm -rf {}" ;'
```
[the remark this answers is not in the document]
Comment (ntdat812) [reply tests/hooks/gateguard-fact-force.test.js]: I agree the file is too long — it was 2,638 lines before this branch touched it — but I'd rather not do the split inside a security fix.

Moving ~2,600 lines of matrices and the shared `runHook`/`runBashHook`/`clearState` harness into new modules would bury a diff whose point is that four specific command spellings flipped from allowed to denied. A reviewer checking that would have to separate the move from the change. It also risks a silent coverage loss: `tests/run-all.js` totals the `Passed:`/`Failed:` lines per file, so a case dropped in the move goes unnoticed unless someone compares counts by hand.

The additions here are 23 lines of table entries in matrices that already existed, not new structure.

Happy to send the split as its own PR — pure move, no behaviour change, with the before/after case count stated so it can be checked — if you'd like it.

**Comment by chenhz01** [comment]
The `1e2aa07` fixes cover the clustered `-c` spellings and value-consuming wrapper options well. One wrapper class still missing from the table: `timeout` (and `setsid`) — `timeout 5m rm -rf /data` reaches the detectors as an invocation of `timeout`, and its first operand is a positional duration, not a flag, so the existing option-arity logic skips it. Same shape as `nice -n`: a wrapper that consumes a positional value before the payload. Adding it to the wrapper set with a positional-arity entry:
```js
{ name: 'timeout', positional: 1 }
```
keeps `sudo`/`timeout`/`env` compositions from re-opening the exact bypass this PR closes. Available as a paid engagement too, if a delivered PR beats a review comment for you.
